# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()
path = "/Users/salizada/Desktop/Springboard/Capstone Project 1/Step 2"
os.chdir(path) 

In [4]:
df_1= pd.read_csv(path +  '/data/step3_output.csv')
df.head()

Unnamed: 0  Unnamed: 0.1                 Name   Region    state  \
0           0             0       Alyeska Resort   Alaska   Alaska   
1           1             1  Eaglecrest Ski Area   Alaska   Alaska   
2           2             2     Hilltop Ski Area   Alaska   Alaska   
3           3             3     Arizona Snowbowl  Arizona  Arizona   
4           4             4  Sunrise Park Resort  Arizona  Arizona   

   summit_elev  vertical_drop  base_elev  trams  fastEight  ...  \
0         3939           2500        250      1        0.0  ...   
1         2600           1540       1200      0        0.0  ...   
2         2090            294       1796      0        0.0  ...   
3        11500           2300       9200      0        0.0  ...   
4        11100           1800       9200      0        NaN  ...   

   LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  \
0            1.0             1610.0           113.0             150.0   
1            2.0              640.0            60.0              45.0   
2            1.0               30.0            30.0             150.0   
3            2.0              777.0           104.0             122.0   
4            1.2              800.0            80.0             115.0   

   yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0       60.0            669.0          65.0          85.0              150.0   
1       44.0            350.0          47.0          53.0               90.0   
2       36.0             69.0          30.0          34.0              152.0   
3       81.0            260.0          89.0          89.0              122.0   
4       49.0            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             NaN  
2            30.0  
3             NaN  
4            80.0  

[5 rows x 29 columns]

In [5]:
df_1.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Name', 'Region', 'state', 'summit_elev',
       'vertical_drop', 'base_elev', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs',
       'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac',
       'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall',
       'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac'],
      dtype='object')

In [6]:
df_1= df_1.drop(columns= ['Unnamed: 0'])

In [7]:
df_1= df_1.drop(columns= ['Unnamed: 0.1'])

In [8]:
df_1.columns

Index(['Name', 'Region', 'state', 'summit_elev', 'vertical_drop', 'base_elev',
       'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac'],
      dtype='object')

In [9]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   Region             330 non-null    object 
 2   state              330 non-null    object 
 3   summit_elev        330 non-null    int64  
 4   vertical_drop      330 non-null    int64  
 5   base_elev          330 non-null    int64  
 6   trams              330 non-null    int64  
 7   fastEight          164 non-null    float64
 8   fastSixes          330 non-null    int64  
 9   fastQuads          330 non-null    int64  
 10  quad               330 non-null    int64  
 11  triple             330 non-null    int64  
 12  double             330 non-null    int64  
 13  surface            330 non-null    int64  
 14  total_chairs       330 non-null    int64  
 15  Runs               326 non-null    float64
 16  TerrainParks       279 non

In [10]:
df_1= df_1.drop(columns=['Region'], axis=1)

In [11]:
df_1.columns

Index(['Name', 'state', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac'],
      dtype='object')

In [37]:
df_1.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0        0.0          0          2     2  ...            1.0   
1        0.0          0          0     0  ...            2.0   
2        0.0          0          0     0  ...            1.0   
3        0.0          1          0     2  ...            2.0   
4        NaN          0          1     2  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             NaN  
2            30.0  
3             NaN  
4            80.0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [12]:
df = pd.concat([df_1.drop('state', axis=1), pd.get_dummies(df_1["state"])], axis=1)

In [13]:
df.shape

(330, 60)

In [18]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        NaN          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [25]:
df=df.dropna()
df.head(10)

Name  summit_elev  vertical_drop  base_elev  trams  \
0          Alyeska Resort         3939           2500        250      1   
2        Hilltop Ski Area         2090            294       1796      0   
41       Ski Granby Ranch         9202           1000       8202      0   
50      Powder Ridge Park          720            550        170      0   
51            Ski Sundown         1075            625        450      0   
58  Pebble Creek Ski Area         8560           2200       6360      0   
61        Silver Mountain         6300           2200       4100      1   
69            Paoli Peaks          900            300        600      0   
72             Seven Oaks          975            275        800      0   
82           Sunday River         3140           2340        800      1   

    fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0         0.0          0          2     2       0  ...             0   
2         0.0          0          0     0       1  ...             0   
41        0.0          0          2     0       1  ...             0   
50        0.0          0          0     0       1  ...             0   
51        0.0          0          0     0       3  ...             0   
58        0.0          0          0     0       3  ...             0   
61        0.0          0          0     1       2  ...             0   
69        0.0          0          1     1       3  ...             0   
72        0.0          0          0     0       2  ...             0   
82        0.0          0          4     5       3  ...             0   

    South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0              0          0     0        0         0           0   
2              0          0     0        0         0           0   
41             0          0     0        0         0           0   
50             0          0     0        0         0           0   
51             0          0     0        0         0           0   
58             0          0     0        0         0           0   
61             0          0     0        0         0           0   
69             0          0     0        0         0           0   
72             0          0     0        0         0           0   
82             0          0     0        0         0           0   

    West Virginia  Wisconsin  Wyoming  
0               0          0        0  
2               0          0        0  
41              0          0        0  
50              0          0        0  
51              0          0        0  
58              0          0        0  
61              0          0        0  
69              0          0        0  
72              0          0        0  
82              0          0        0  

[10 rows x 60 columns]

In [26]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [27]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [28]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [29]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred= model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [30]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
mean_absolute_error(y_test, y_pred)

8.358680279458271

In [31]:
explained_variance_score(y_test, y_pred)

0.7069799328598516

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [32]:
lm.intercept_

61.18936167993579

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [33]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
LongestRun_mi      1.012657e+01
AdultWeekday       9.780789e+00
Montana            8.292979e+00
TerrainParks       8.233056e+00
averageSnowfall    7.480548e+00
SkiableTerrain_ac  5.786552e+00
projectedDaysOpen  5.541603e+00
North Carolina     4.852197e+00
Idaho              4.707552e+00
surface            4.616947e+00
quad               3.748302e+00
vertical_drop      3.429838e+00
New Mexico         3.427162e+00
trams              3.381248e+00
Pennsylvania       3.214633e+00
Snow Making_ac     3.154865e+00
NightSkiing_ac     3.042541e+00
daysOpenLastYear   2.466544e+00
Wisconsin          2.431037e+00
triple             2.230925e+00
Massachusetts      2.129126e+00
Minnesota          2.027542e+00
Maine              1.966413e+00
Michigan           1.880289e+00
yearsOpen          1.844931e+00
Indiana            1.805989e+00
summit_elev        1.739070e+00
Runs               1.609544e+00
New York           1.350939e+00
fastSixes          1.195983e+00
Connecticut        1.080997e+00
Alaska             1.046397e+00
fastQuads          1.033272e+00
Iowa               8.153480e-01
total_chairs       7.917945e-01
double             5.933427e-01
Vermont            5.017717e-01
base_elev          2.617490e-01
Utah               2.057542e-01
New Hampshire      1.955109e-14
Colorado           7.320048e-15
California         4.329870e-15
Arizona            3.552714e-15
Illinois           8.881784e-16
fastEight          4.440892e-16
New Jersey         0.000000e+00
Ohio               0.000000e+00
Oregon             0.000000e+00
Nevada             0.000000e+00
Rhode Island       0.000000e+00
South Dakota       0.000000e+00
Tennessee          0.000000e+00
Missouri           0.000000e+00
Maryland           0.000000e+00
Virginia           0.000000e+00
Washington         0.000000e+00
West Virginia      0.000000e+00
Wyoming            0.000000e+00

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [105]:
df_2= df_1.copy()

In [106]:
df_2=df_2.dropna()

In [107]:
from sklearn import preprocessing

X_1 = df_2.drop(['Name','state','AdultWeekend'], axis=1)

y_1 = df_2.AdultWeekend

scaler_1 = preprocessing.StandardScaler().fit(X_1)

X1_scaled=scaler_1.transform(X_1) 

In [108]:
from sklearn.model_selection import train_test_split

y_1 = y_1.ravel()
 
X_1_train, X_1_test, y_1_train, y_1_test= train_test_split(X1_scaled, y_1, test_size=0.25, random_state=1)

In [114]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm1 = linear_model.LinearRegression()
model_1 = lm1.fit(X_1_train,y_1_train)

In [115]:
y_1_pred= model_1.predict(X_1_test)

In [116]:
mean_absolute_error(y_1_test, y_1_pred)

21.67192683732176

In [117]:
explained_variance_score(y_1_test, y_1_pred)

-1.9674206738272217

In [118]:
lm1.intercept_

58.737182015203786

In [119]:
pd.DataFrame(abs(lm1.coef_), X_1.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
summit_elev        1.265880e+02
base_elev          1.060039e+02
vertical_drop      2.163461e+01
SkiableTerrain_ac  2.031425e+01
projectedDaysOpen  1.313311e+01
daysOpenLastYear   1.113992e+01
TerrainParks       8.914323e+00
quad               6.874756e+00
Runs               6.256042e+00
surface            5.834203e+00
AdultWeekday       4.343705e+00
LongestRun_mi      4.323448e+00
NightSkiing_ac     2.663881e+00
double             2.384410e+00
fastSixes          2.103262e+00
total_chairs       1.152987e+00
triple             1.112027e+00
trams              9.957838e-01
averageSnowfall    8.041857e-01
yearsOpen          6.075963e-01
fastQuads          1.849201e-01
Snow Making_ac     1.392505e-01
fastEight          2.557954e-13

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [76]:
df_3= df_1.copy()

In [77]:
df_3=df_3.dropna()

In [78]:
df_3.columns

Index(['Name', 'state', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac'],
      dtype='object')

In [120]:
from sklearn import preprocessing

X_2 = df_3.drop(['Name','state','AdultWeekend','summit_elev','base_elev'], axis=1)

y_2 = df_3.AdultWeekend

scaler_2 = preprocessing.StandardScaler().fit(X_2)

X2_scaled=scaler_2.transform(X_2)

In [121]:
from sklearn.model_selection import train_test_split

y_2 = y_2.ravel()
 
X_2_train, X_2_test, y_2_train, y_2_test= train_test_split(X2_scaled, y_2, test_size=0.25, random_state=1)

In [122]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm2 = linear_model.LinearRegression()
model_2 = lm2.fit(X_2_train,y_2_train)

In [123]:
y_2_pred= model_2.predict(X_2_test)

In [124]:
print(mean_absolute_error(y_2_test, y_2_pred))
print(explained_variance_score(y_2_test, y_2_pred))
print(lm2.intercept_)

21.036036098663953
-1.6543061186216206
58.766937605292384


In [86]:
pd.DataFrame(abs(lm.coef_), X_2.columns, columns=['Coefficient']).sort_values(by=['Coefficient'], ascending=False)

Coefficient
SkiableTerrain_ac  1.903567e+01
projectedDaysOpen  1.185957e+01
vertical_drop      1.066546e+01
daysOpenLastYear   1.001598e+01
TerrainParks       8.850045e+00
Snow Making_ac     8.660351e+00
Runs               7.267809e+00
AdultWeekday       5.912903e+00
surface            5.877445e+00
quad               5.468478e+00
LongestRun_mi      5.357030e+00
averageSnowfall    4.353034e+00
fastQuads          3.917217e+00
trams              3.169738e+00
fastSixes          3.068960e+00
double             2.688365e+00
NightSkiing_ac     2.627769e+00
total_chairs       2.004701e+00
triple             1.757074e+00
yearsOpen          1.099139e+00
fastEight          4.440892e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.7069799328598516 | 8.358680279458271 |-|
| Model 2. | -1.9674206738272217|21.67192683732176 |'state'|
| Model 3. | -1.6543061186216206 | 21.0360360986639531|'state','summit_elev','base_elev'|

Model 1 has been selected for the success, because: 1. MAE=<40 and 2. EV is closer to 1